crates/biome_js_analyze/src/lint/nursery/no_flat_map_identity.rs

```diff
@@ -0,0 +1,141 @@
+use biome_js_syntax::{
+    AnyJsExpression, AnyJsFunctionBody, AnyJsMemberExpression, AnyJsName, AnyJsStatement,
+    JsCallExpression, JsSyntaxKind,
+};
+use biome_rowan::{AstNode, AstSeparatedList, BatchMutationExt};
+
+use crate::JsRuleAction;
+
+declare_rule! {
+    /// Disallow to use unnecessary callback on `flatMap`.
+    ///
+    /// To achieve the same result (flattening an array) more concisely and efficiently, you should use `flat` instead.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// array.flatMap((arr) => arr);
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// array.flatMap((arr) => {return arr});
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// array.flatMap((arr) => arr * 2);
+    /// ```
+    ///
+    pub NoFlatMapIdentity {
+        version: "next",
+        name: "noFlatMapIdentity",
+        recommended: true,
+        source: RuleSource::Clippy("flat_map_identity"),
+        fix_kind: FixKind::Safe,
+    }
+}
+
+impl Rule for NoFlatMapIdentity {
+    type Query = Ast<JsCallExpression>;
+    type State = ();
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let flat_map_call = ctx.query();
+        let arguments = flat_map_call.arguments().ok()?.args();
+        if let Some(arg) = arguments.first() {
+            let arg = arg.ok()?;
+            let arg = arg
+                .as_any_js_expression()?
+                .as_js_arrow_function_expression()?;
+
+            let function_parameter = arg.parameters().ok()?.text();
+
+            let function_body: String = match arg.body().ok()? {
+                AnyJsFunctionBody::AnyJsExpression(body) => body.omit_parentheses().text(),
+                AnyJsFunctionBody::JsFunctionBody(body) => {
+                    let mut statement = body.statements().into_iter();
+                    match statement.next() {
+                        Some(AnyJsStatement::JsReturnStatement(body)) => {
+                            let Some(AnyJsExpression::JsIdentifierExpression(return_statement)) =
+                                body.argument()
+                            else {
+                                return None;
+                            };
+                            return_statement.name().ok()?.text()
+                        }
+                        _ => return None,
+                    }
+                }
+            };
+
+            let function_parameter = function_parameter.trim_matches(&['(', ')']);
+            if function_parameter != function_body {
+                return None;
+            }
+            let flat_map_expression =
+                AnyJsMemberExpression::cast_ref(flat_map_call.callee().ok()?.syntax())?;
+            if flat_map_expression.member_name()?.text() != "flatMap" {
+                return None;
+            }
```

Comment:
This should be checked first, before `if let Some(arg) = arguments.first()`, or every function call with zero arguments will trigger this rule.

Reply:
done